app/controllers/common/PaymentDetailsController.scala

```diff
@@ -0,0 +1,81 @@
+import controllers.common.PaymentDetailsController.endOfPaymentDate
+import forms.common.PaymentDetailsFormProvider
+import models.enumeration.EventType
+import models.{Index, TaxYear, UserAnswers}
+import pages.common.PaymentDetailsPage
+import pages.{TaxYearPage, Waypoints}
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.common.PaymentDetailsView
+
+import java.time.{LocalDate, Month}
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class PaymentDetailsController @Inject()(val controllerComponents: MessagesControllerComponents,
+                                         identify: IdentifierAction,
+                                         getData: DataRetrievalAction,
+                                         userAnswersCacheConnector: UserAnswersCacheConnector,
+                                         formProvider: PaymentDetailsFormProvider,
+                                         view: PaymentDetailsView
+                                        )(implicit ec: ExecutionContext) extends FrontendBaseController with I18nSupport { // scalastyle:off magic.number
+
+  def onPageLoad(waypoints: Waypoints, eventType: EventType, index: Index): Action[AnyContent] = (identify andThen getData(eventType)) { implicit request =>
+    val startDate = LocalDate.of(2006, Month.APRIL, 6)
+    val endDate = endOfPaymentDate(request.userAnswers.flatMap(_.get(TaxYearPage)))
+
+    val preparedForm = request.userAnswers.flatMap(_.get(PaymentDetailsPage(eventType, index))) match {
+      case Some(value) => formProvider(startDate, endDate).fill(value)
+      case None => formProvider(startDate, endDate)
+    }
+    Ok(view(preparedForm, waypoints, eventType, index))
+  }
+
+  def onSubmit(waypoints: Waypoints, eventType: EventType, index: Index): Action[AnyContent] = (identify andThen getData(eventType)).async { implicit request =>
+    val startDate = LocalDate.of(2006, Month.APRIL, 6)
+    val endDate = endOfPaymentDate(request.userAnswers.flatMap(_.get(TaxYearPage)))
+
+    formProvider(startDate, endDate).bindFromRequest().fold(
+      formWithErrors => {
+        Future.successful(BadRequest(view(formWithErrors, waypoints, eventType, index)))
+      },
+      value => {
+        val originalUserAnswers = request.userAnswers.fold(UserAnswers())(identity)
+        val updatedAnswers = originalUserAnswers.setOrException(PaymentDetailsPage(eventType, index), value)
+        userAnswersCacheConnector.save(request.pstr, eventType, updatedAnswers).map { _ =>
+          Redirect(PaymentDetailsPage(eventType, index).navigate(waypoints, originalUserAnswers, updatedAnswers).route)
+        }
+      }
+    )
+  }
+}
+
+object PaymentDetailsController {
+  private def endOfPaymentDate(optTaxYear: Option[TaxYear]): LocalDate = optTaxYear match {
```

Comment:
Can we use TaxYear.getSelectedTaxYearAsString instead? We don't want to proceed if tax year is unavailable. throwing an exception here is fine.

Reply:
Done